### PR TITLE
Add set_freezable() for per-object freeze policy

### PR DESCRIPTION
**Pull request by mjp41**
Add `set_freezable(obj, status)` to control whether individual objects
can be frozen. The status is one of four values:

  `FREEZABLE_YES`      - always freezable
  `FREEZABLE_NO`       - never freezable
  `FREEZABLE_EXPLICIT` - freezable only when freeze() is called directly
                       on the object, not when reached as a child
  `FREEZABLE_PROXY`    - reserved for future module proxy use

Storage uses a two-tier strategy: first tries setting a `__freezable__`
attribute on the object (works for any object with a `__dict__`), falling
back to a weakref-keyed dictionary in the interpreter state (works for
`__slots__` objects with `__weakref__` support). If neither is possible,
raises TypeError.

During freeze, `check_freezable()` queries the status by checking the
`__freezable__` attribute first, then the weakref dictionary. For
EXPLICIT, the freeze root object is tracked in `FreezeState` so that
direct vs child freezing can be distinguished.

The old standalone `__freezable__ = False` check in `traverse_freeze` is
removed; all `__freezable__` handling is now unified through
`check_freezable()` and `_PyImmutability_GetFreezable()`.

Changes:
- `Include/internal/pycore_immutability.h`: add `_Py_freezable_status`
  enum, `freezable_objects` dict, and `destroy_objects_cb` to state
- `Include/cpython/immutability.h`: declare `SetFreezable`/`GetFreezable`
- `Python/immutability.c`: implement `SetFreezable` (attr-first, weakref
  fallback), `GetFreezable` (attr-first, weakref fallback),
  `_destroy_dict` callback, `FreezeState.root`, updated `check_freezable`
- `Modules/_immutablemodule.c`: expose `set_freezable()` and constants
- `Lib/immutable.py`: export `set_freezable` and `FREEZABLE_*` constants
- `Python/pystate.c`: cleanup new state fields in `interpreter_clear()`
- `Lib/test/test_freeze/test_set_freezable.py`: 18 tests covering all
  status values, storage strategies, edge cases, and constants
- Migrate `_blocking_on` and `_module_locks` to use this mechanism.

### CODE REVIEW

File: Include/internal/pycore_immutability.h [hunk 1]
```diff
@@ -10,13 +10,21 @@ extern "C" {
...
 struct _Py_immutability_state {
-    PyObject *module_locks;
-    PyObject *blocking_on;
+    int late_init_done;
     PyObject *freezable_types;
     _Py_hashtable_t *shallow_immutable_types;
     PyObject *destroy_cb;
     _Py_hashtable_t *warned_types;
+    PyObject *freezable_objects;      // dict: weakref(obj) -> int status
+    PyObject *destroy_objects_cb;     // weak-ref callback for freezable_objects
```
Comment (xFrednet) [inline Include/internal/pycore_immutability.h]: Is the usage of weakrefs safe here?

File: Lib/test/test_freeze/test_set_freezable.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        set_freezable(obj, FREEZABLE_YES)
+        freeze(obj)
+        self.assertTrue(isfrozen(obj))
+
+    def test_freeze_as_child_succeeds(self):
+        C = make_freezable_class()
+        parent = C()
+        child = C()
+        parent.child = child
+        set_freezable(child, FREEZABLE_YES)
```
Comment (xFrednet) [inline Lib/test/test_freeze/test_set_freezable.py:33]:
```suggestion
```

shouldn't be needed

File: Lib/test/test_freeze/test_set_freezable.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        self.assertFalse(isfrozen(obj))
+
+    def test_freeze_as_child_raises(self):
+        C = make_freezable_class()
+        parent = C()
+        child = C()
+        parent.child = child
+        set_freezable(child, FREEZABLE_NO)
+        with self.assertRaises(TypeError):
+            freeze(parent)
```
Comment (xFrednet) [inline Lib/test/test_freeze/test_set_freezable.py:56]:
```suggestion
            freeze(parent)
        self.assertFalse(isfrozen(child))
        self.assertFalse(isfrozen(parent))
```

File: Lib/test/test_freeze/test_set_freezable.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        self.assertFalse(isfrozen(child))
+
+
+class TestSetFreezableProxy(unittest.TestCase):
+    """FREEZABLE_PROXY: reserved — currently falls through to default behaviour."""
+
+    def test_proxy_allows_freeze(self):
+        C = make_freezable_class()
+        obj = C()
+        set_freezable(obj, FREEZABLE_PROXY)
```
Comment (xFrednet) [inline Lib/test/test_freeze/test_set_freezable.py]: Proxy should only be allowed for modules. I think this should already throw a type error

File: Lib/test/test_freeze/test_set_freezable.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        obj = C()
+        with self.assertRaises(ValueError):
+            set_freezable(obj, 99)
+        with self.assertRaises(ValueError):
+            set_freezable(obj, -1)
+
+    def test_object_without_weakref_support_raises(self):
+        # Built-in ints don't support weak references or attribute setting.
+        with self.assertRaises(TypeError):
+            set_freezable(42, FREEZABLE_NO)
```
Comment (xFrednet) [inline Lib/test/test_freeze/test_set_freezable.py]: Should we allow setting freezability for shallow immutable objects in the first place?

File: Python/immutability.c [hunk 6]
```diff
@@ -1243,10 +1235,30 @@ is_explicitly_freezable(struct _Py_immutability_state *state, PyObject *obj)
...
+            goto error;
+        case _Py_FREEZABLE_PROXY:
+            // Reserved for future use — fall through to existing checks.
+            break;
+        }
+    }
+
     // Check is object is subclass of NotFreezable
     // TODO: Would be nice for this to be faster.
     if (PyObject_IsInstance(obj, (PyObject *)&_PyNotFreezable_Type) == 1){
```
Comment (xFrednet) [inline Python/immutability.c:1238]: Now we should be able to remove the `_PyNotFreezable_Type` (Can be done in a follow up.)

File: Python/immutability.c [hunk 7]
```diff
@@ -1651,6 +1770,58 @@ static int traverse_freeze(PyObject* obj, struct FreezeState* freeze_state)
     return -1;
 }
 
+// Mark importlib's mutable state as not freezable.
+// Separated from init_state because _frozen_importlib is not
+// available during early interpreter startup.
+static void
+late_init(struct _Py_immutability_state *state)
```
Comment (xFrednet) [inline Python/immutability.c:1748]: Excellent! :D

**Review by xFrednet — COMMENTED** [review]
Some high level nits, but the rest LGTM. Nice work :D

File: Lib/test/test_freeze/test_set_freezable.py (hunk 5 above)
Comment (mjp41) [reply Lib/test/test_freeze/test_set_freezable.py]: Good question, I don't know.  There are a bunch of defaults I am not sure on the right choices for.

File: Include/internal/pycore_immutability.h (hunk 1 above)
Comment (mjp41) [reply Include/internal/pycore_immutability.h]: Moved to using ob_flags as discussed on Teams.
